#Set the working directory


#Read in all libraries


#Source in all other scripts and/or Rdata files

#######################
#Within each subpopulation
#############
#Run a GWAS using SNPs


#Filter out the GWAS results 


#Calculate pearson correlations between SNPs additive effects that are in the 
#vicinity of core genes

#Calculate pearson correlations between SNP additive effects that are in the 
# vicinity of peripheral genes


#############
#Run a GWAS using QTNs



#Filter out the GWAS results 


#Calculate pearson correlations between QTN additive effects that are in the 
#vicinity of core genes

#Calculate pearson correlations between SNP additive effects that are in the 
# vicinity of peripheral genes